akka-http-core/src/main/scala/akka/http/impl/engine/client/ProxyGraphStage.scala

```diff
@@ -0,0 +1,91 @@
+package akka.http.impl.engine.client
+
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+import akka.stream.{ Attributes, BidiShape, Inlet, Outlet }
+import akka.util.ByteString
+
+class ProxyGraphStage(targetHostName: String) extends GraphStage[BidiShape[ByteString, ByteString, ByteString, ByteString]] {
+  sealed trait State
+  case object Starting extends State
+  case class Connecting(waitingElem: ByteString) extends State
+  case object Connected extends State
+
+  val bytesIn: Inlet[ByteString] = Inlet("OutgoingTCP.in")
+  val bytesOut: Outlet[ByteString] = Outlet("OutgoingTCP.out")
+
+  val sslIn: Inlet[ByteString] = Inlet("OutgoingSSL.in")
+  val sslOut: Outlet[ByteString] = Outlet("OutgoingSSL.out")
+
+  override def shape: BidiShape[ByteString, ByteString, ByteString, ByteString] = BidiShape.apply(sslIn, bytesOut, bytesIn, sslOut)
+
+  private val connectMsg = ByteString(s"CONNECT ${targetHostName}:443 HTTP/1.1\r\nHost: ${targetHostName}\r\n\r\n")
+
+  @scala.throws[Exception](classOf[Exception])
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var state: State = Starting
+    private var msgs = List.empty[ByteString]
+
+    setHandler(sslIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting ⇒
+            push(bytesOut, connectMsg)
+            state = Connecting(grab(sslIn))
+          case _: Connecting ⇒
+            msgs = msgs ++ List(grab(sslIn))
+          case Connected ⇒
+            if (msgs.size > 0) {
+              emitMultiple(bytesOut, msgs ++ List(grab(sslIn)))
+              msgs = List.empty
+            } else {
+              push(bytesOut, grab(sslIn))
+            }
+        }
+      }
+    })
+
+    setHandler(bytesIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting ⇒
+          // should not happen to be in Starting state
+          case Connecting(waitingElem) ⇒
+            val el = grab(bytesIn)
+            push(bytesOut, waitingElem)
+            state = Connected
+            pull(bytesIn)
+          case Connected ⇒
+            push(sslOut, grab(bytesIn))
+        }
+      }
+    })
+
+    setHandler(bytesOut, new OutHandler {
+      override def onPull() = {
+        // TODO: rethink
+        state match {
+          case Starting ⇒
+            pullIfNeeded(sslIn)
```

Comment:
Instead of collecting outgoing messages, you shouldn't pull from the user-level layer (`sslIn`) until the connection is established. You can just ignore the pull here and then when the proxy connection has been established, check `isAvailable(bytesOut)` and then `pull(sslIn)`..